sharedbuffers/mapped_struct.py

```diff
@@ -320,6 +320,342 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
+                raise NotImplementedError
+            return proxied_list_cmp(self, other)
+
+        def _ne(self, other):
+            if not islist(other):
+                return True
+            return proxied_list_cmp(self, other) != 0
+
+        def _eq(self, other):
+            if not islist(other):
+                return False
+            return proxied_list_cmp(self, other) == 0
+
+    @cython.ccall
+    @cython.locals(offs = cython.ulonglong, dcode = cython.char, pbuf = 'const char *', objlen = cython.ulonglong)
+    def _cobjlen(self):
+
+        offs = self.offs
+        pbuf = cython.cast(cython.p_char, self.pybuf.buf)
+        dcode = pbuf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen = cython.cast(cython.p_uint, pbuf + offs)[0] >> 8
+            if objlen == 0xFFFFFF:
+                objlen = cython.cast(cython.p_longlong, pbuf + offs + 4)[0]
+
+        elif dcode in ('q', 'd', 't'):
+            objlen = cython.cast(cython.p_longlong, pbuf + offs)[0] >> 8
+
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
+
+    def _objlen(self):
+        offs = self.offs
+
+        buf = self.buf
+        dcode = buf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen, = struct.unpack('<I', buf[offs+1:offs+4] + '\x00')
+            if objlen == 0xFFFFFF:
+                objlen = struct.unpack_from('<Q', buf, offs + 4)
+
+        elif dcode in ('q', 'd', 't'):
+            objlen, = struct.unpack('<Q', buf[offs+1:offs+8] + '\x00')
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
+
+    def __len__(self):
+        if cython.compiled:
+            return self._cobjlen()
+        else:
+            return self._objlen()
+
+    def __nonzero__(self):
+        return len(self) > 0
+
+    def __bool__(self):
+        return len(self) > 0
+
+    def __setitem__(self, index, value):
+        raise AttributeError("Proxy objects are read-only")
+
+    def __delitem__(self, index):
+        raise AttributeError("Proxy objects are read-only")
+
+    def __iter__(self):
+        for i in xrange(len(self)):
+            yield self[i]
+
+    def __reversed__(self):
+        for i in reversed(xrange(len(self))):
+            yield self[i]
+
+    def __contains__(self, item):
+        for e in self:
+            if e ==  item:
+                return True
+        return False
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        return "proxied_list([%s])" % ",".join([str(x) for x in self])
+
+if not cython.compiled:
+    setattr(proxied_list, '__eq__', getattr(proxied_list, '_eq'))
+    setattr(proxied_list, '__ne__', getattr(proxied_list, '_ne'))
+
+@cython.cclass
+class proxied_tuple(proxied_list):
+
+    def __str__(self):
+        return "proxied_tuple([%s])" % ",".join([str(x) for x in self])
```

Comment:
I'm pretty sure you need to implement more stuff for it to be a tuple-like thing.

You need a `__hash__` method for instance.